// testcafe/lib/utils/parse-file-list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const pinkie_1 = __importDefault(require("pinkie"));
const globby_1 = __importDefault(require("globby"));
const is_glob_1 = __importDefault(require("is-glob"));
const compiler_1 = __importDefault(require("../compiler"));
const os_family_1 = __importDefault(require("os-family"));
const lodash_1 = require("lodash");
const promisified_functions_1 = require("../utils/promisified-functions");
const DEFAULT_TEST_LOOKUP_DIRS = ['test/', 'tests/'];
const TEST_FILE_GLOB_PATTERN = `./**/*@(${compiler_1.default.getSupportedTestFileExtensions().join('|')})`;
function modifyFileRoot(baseDir, file) {
    const absPath = path_1.default.resolve(baseDir, file);
    const fileIsOnOtherDrive = path_1.default.isAbsolute(path_1.default.relative(baseDir, file));
    if (!path_1.default.isAbsolute(file) || fileIsOnOtherDrive)
        return file;
    const { root, dir, base } = path_1.default.parse(absPath);
    return path_1.default.join(path_1.default.parse(baseDir).root, path_1.default.relative(root, dir), base);
}
async function getDefaultDirs(baseDir) {
    return await globby_1.default(DEFAULT_TEST_LOOKUP_DIRS, {
        cwd: baseDir,
        nocase: true,
        silent: true
    });
}
async function convertDirsToGlobs(fileList, baseDir) {
    fileList = await pinkie_1.default.all(fileList.map(async (file) => {
        if (!is_glob_1.default(file)) {
            const absPath = path_1.default.resolve(baseDir, file);
            let fileStat = null;
            try {
                fileStat = await promisified_functions_1.stat(absPath);
            }
            catch (err) {
                return null;
            }
            if (fileStat.isDirectory())
                return path_1.default.join(file, TEST_FILE_GLOB_PATTERN);
            if (os_family_1.default.win)
                file = modifyFileRoot(baseDir, file);
        }
        return file;
    }));
    return fileList.filter(file => !!file);
}
async function parseFileList(fileList, baseDir) {
    if (lodash_1.isEmpty(fileList))
        fileList = await getDefaultDirs(baseDir);
    fileList = await convertDirsToGlobs(fileList, baseDir);
    fileList = await globby_1.default(fileList, { cwd: baseDir });
    return fileList.map(file => path_1.default.resolve(baseDir, file));
}
exports.default = parseFileList;
module.exports = exports.default;
