// testcafe/lib/utils/promisified-functions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = __importDefault(require("child_process"));
const graceful_fs_1 = __importDefault(require("graceful-fs"));
const pinkie_1 = __importDefault(require("pinkie"));
const pngjs_1 = require("pngjs");
const promisify_event_1 = __importDefault(require("promisify-event"));
const promisify_1 = __importDefault(require("./promisify"));
exports.readDir = promisify_1.default(graceful_fs_1.default.readdir);
exports.stat = promisify_1.default(graceful_fs_1.default.stat);
exports.writeFile = promisify_1.default(graceful_fs_1.default.writeFile);
exports.readFile = promisify_1.default(graceful_fs_1.default.readFile);
exports.deleteFile = promisify_1.default(graceful_fs_1.default.unlink);
exports.exec = promisify_1.default(child_process_1.default.exec);
exports.sendMessageToChildProcess = promisify_1.default((process, ...args) => process.send(...args));
function readPng(buffer) {
    const png = new pngjs_1.PNG();
    const parsedPromise = pinkie_1.default.race([
        promisify_event_1.default(png, 'parsed'),
        promisify_event_1.default(png, 'error')
    ]);
    png.parse(buffer);
    return parsedPromise
        .then(() => png);
}
exports.readPng = readPng;
async function readPngFile(filePath) {
    const buffer = await exports.readFile(filePath);
    return await readPng(buffer);
}
exports.readPngFile = readPngFile;
function writePng(filePath, png) {
    const outStream = graceful_fs_1.default.createWriteStream(filePath);
    const pngStream = png.pack();
    const finishPromise = pinkie_1.default.race([
        promisify_event_1.default(outStream, 'finish'),
        promisify_event_1.default(outStream, 'error'),
        promisify_event_1.default(pngStream, 'error')
    ]);
    pngStream.pipe(outStream);
    return finishPromise;
}
exports.writePng = writePng;
