// testcafe/lib/browser/connection/remotes-queue.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const pinkie_1 = __importDefault(require("pinkie"));
const events_1 = require("events");
const promisify_event_1 = __importDefault(require("promisify-event"));
const time_limit_promise_1 = __importDefault(require("time-limit-promise"));
const REMOTE_REDIRECT_TIMEOUT = 10000;
const ADDING_CONNECTION_WAITING_TIMEOUT = 10000;
class RemotesQueue {
    constructor() {
        this.events = new events_1.EventEmitter();
        this.shiftingTimeout = pinkie_1.default.resolve();
        this.pendingConnections = {};
    }
    add(remoteConnection) {
        const connectionReadyPromise = promisify_event_1.default(remoteConnection, 'ready')
            .then(() => this.remove(remoteConnection));
        this.pendingConnections[remoteConnection.id] = {
            connection: remoteConnection,
            readyPromise: connectionReadyPromise
        };
        this.events.emit('connection-added', remoteConnection.id);
    }
    remove(remoteConnection) {
        delete this.pendingConnections[remoteConnection.id];
    }
    shift() {
        const shiftingPromise = this.shiftingTimeout
            .then(async () => {
            let headId = Object.keys(this.pendingConnections)[0];
            if (!headId)
                headId = await time_limit_promise_1.default(promisify_event_1.default(this.events, 'connection-added'), ADDING_CONNECTION_WAITING_TIMEOUT);
            return headId ? this.pendingConnections[headId].connection : null;
        });
        this.shiftingTimeout = shiftingPromise
            .then(connection => {
            if (!connection)
                return pinkie_1.default.resolve();
            return time_limit_promise_1.default(this.pendingConnections[connection.id].readyPromise, REMOTE_REDIRECT_TIMEOUT);
        });
        return shiftingPromise;
    }
}
exports.default = RemotesQueue;
module.exports = exports.default;
