// testcafe/lib/runner/tested-app.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
const path_1 = require("path");
const pinkie_1 = __importDefault(require("pinkie"));
const tree_kill_1 = __importDefault(require("tree-kill"));
const os_family_1 = __importDefault(require("os-family"));
const delay_1 = __importDefault(require("../utils/delay"));
const runtime_1 = require("../errors/runtime");
const types_1 = require("../errors/types");
const resolve_path_relatively_cwd_1 = __importDefault(require("../utils/resolve-path-relatively-cwd"));
const MODULES_BIN_DIR = resolve_path_relatively_cwd_1.default('./node_modules/.bin');
const ENV_PATH_KEY = (function () {
    if (os_family_1.default.win) {
        let pathKey = 'Path';
        Object.keys(process.env).forEach(key => {
            if (key.toLowerCase() === 'path')
                pathKey = key;
        });
        return pathKey;
    }
    return 'PATH';
})();
class TestedApp {
    constructor() {
        this.process = null;
        this.errorPromise = null;
        this.killed = false;
    }
    async start(command, initDelay) {
        this.errorPromise = new pinkie_1.default((resolve, reject) => {
            const env = Object.assign({}, process.env);
            const path = env[ENV_PATH_KEY] || '';
            const pathParts = path.split(path_1.delimiter);
            pathParts.unshift(MODULES_BIN_DIR);
            env[ENV_PATH_KEY] = pathParts.join(path_1.delimiter);
            this.process = child_process_1.exec(command, { env }, err => {
                if (!this.killed && err) {
                    const message = err.stack || String(err);
                    reject(new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.testedAppFailedWithError, message));
                }
            });
        });
        await pinkie_1.default.race([
            delay_1.default(initDelay),
            this.errorPromise
        ]);
    }
    async kill() {
        this.killed = true;
        const killPromise = new pinkie_1.default(resolve => tree_kill_1.default(this.process.pid, 'SIGTERM', resolve));
        await killPromise;
    }
}
exports.default = TestedApp;
module.exports = exports.default;
