// testcafe-hammerhead/lib/shadow-ui/create-shadow-stylesheet.js
"use strict";

exports.__esModule = true;
exports.default = createShadowStylesheet;

var _css = _interopRequireDefault(require("css"));

var _className = _interopRequireDefault(require("./class-name"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const ID_OR_CLASS_RE = /#[a-zA-Z0-9_-]+|\.-?[a-zA-Z0-9_][a-zA-Z0-9_-]*/g;
const ADD_POSTFIX_REPLACEMENT = '$&' + _className.default.postfix;

function transformSelector(selector) {
  return selector.replace(ID_OR_CLASS_RE, ADD_POSTFIX_REPLACEMENT);
}

function addUIClassPostfix(rules) {
  let ruleStack = rules.slice();

  while (ruleStack.length) {
    const rule = ruleStack.pop();
    if (rule.type === 'rule' && rule.selectors) rule.selectors = rule.selectors.map(transformSelector);
    if (rule.rules) ruleStack = ruleStack.concat(rule.rules);
  }
}

function createShadowStylesheet(cssCode) {
  const ast = _css.default.parse(cssCode, {
    silent: true
  });

  addUIClassPostfix(ast.stylesheet.rules);
  return _css.default.stringify(ast, {
    indent: 0,
    compress: false
  });
}

module.exports = exports.default;
