// testcafe/lib/browser/provider/built-in/dedicated/base.js
'use strict';

exports.__esModule = true;

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _testcafeBrowserTools = require('testcafe-browser-tools');

var _getMaximizedHeadlessWindowSize = require('../../utils/get-maximized-headless-window-size');

var _getMaximizedHeadlessWindowSize2 = _interopRequireDefault(_getMaximizedHeadlessWindowSize);

var _crop = require('../../../../screenshots/crop');

var _promisifiedFunctions = require('../../../../utils/promisified-functions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    openedBrowsers: {},

    isMultiBrowser: false,

    _getConfig() {
        throw new Error('Not implemented');
    },

    _getBrowserProtocolClient() /* runtimeInfo */{
        throw new Error('Not implemented');
    },

    _getBrowserName() {
        return this.providerName.replace(':', '');
    },

    isValidBrowserName(browserName) {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const config = yield _this._getConfig(browserName);
            const browserInfo = yield (0, _testcafeBrowserTools.getBrowserInfo)(config.path || _this._getBrowserName());

            return !!browserInfo;
        })();
    },

    isLocalBrowser() {
        return (0, _asyncToGenerator3.default)(function* () {
            return true;
        })();
    },

    isHeadlessBrowser(browserId) {
        return this.openedBrowsers[browserId].config.headless;
    },

    _getCropDimensions(viewportWidth, viewportHeight) {
        if (!viewportWidth || !viewportHeight) return null;

        return {
            left: 0,
            top: 0,
            right: viewportWidth,
            bottom: viewportHeight
        };
    },

    takeScreenshot(browserId, path, viewportWidth, viewportHeight) {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const runtimeInfo = _this2.openedBrowsers[browserId];
            const browserClient = _this2._getBrowserProtocolClient(runtimeInfo);
            const binaryImage = yield browserClient.getScreenshotData(runtimeInfo);
            const pngImage = yield (0, _promisifiedFunctions.readPng)(binaryImage);
            const cropDimensions = _this2._getCropDimensions(viewportWidth, viewportHeight);
            const croppedImage = yield (0, _crop.cropScreenshot)(pngImage, { path, cropDimensions });

            yield (0, _promisifiedFunctions.writePng)(path, croppedImage || pngImage);
        })();
    },

    maximizeWindow(browserId) {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const maximumSize = (0, _getMaximizedHeadlessWindowSize2.default)();

            yield _this3.resizeWindow(browserId, maximumSize.width, maximumSize.height, maximumSize.width, maximumSize.height);
        })();
    }
};
module.exports = exports['default'];
