// testcafe/lib/configuration/option-names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    src: 'src',
    browsers: 'browsers',
    concurrency: 'concurrency',
    filter: 'filter',
    filterTestGrep: 'filter.testGrep',
    filterFixtureGrep: 'filter.fixtureGrep',
    reporter: 'reporter',
    ssl: 'ssl',
    speed: 'speed',
    screenshotPath: 'screenshotPath',
    screenshotPathPattern: 'screenshotPathPattern',
    takeScreenshotsOnFails: 'takeScreenshotsOnFails',
    proxyBypass: 'proxyBypass',
    appCommand: 'appCommand',
    appInitDelay: 'appInitDelay',
    skipJsErrors: 'skipJsErrors',
    disablePageReloads: 'disablePageReloads',
    quarantineMode: 'quarantineMode',
    debugMode: 'debugMode',
    debugOnFail: 'debugOnFail',
    skipUncaughtErrors: 'skipUncaughtErrors',
    stopOnFirstFail: 'stopOnFirstFail',
    selectorTimeout: 'selectorTimeout',
    assertionTimeout: 'assertionTimeout',
    pageLoadTimeout: 'pageLoadTimeout',
    videoPath: 'videoPath',
    videoOptions: 'videoOptions',
    videoEncodingOptions: 'videoEncodingOptions'
};
module.exports = exports.default;
