// testcafe/lib/video-recorder/test-run-video-recorder.js
'use strict';

exports.__esModule = true;

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _entries = require('babel-runtime/core-js/object/entries');

var _entries2 = _interopRequireDefault(_entries);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _path = require('path');

var _process = require('./process');

var _process2 = _interopRequireDefault(_process);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const VIDEO_EXTENSION = 'mp4';

const TEMP_VIDEO_FILE_PREFIX = 'tmp-video';
const TEMP_MERGE_FILE_PREFIX = TEMP_VIDEO_FILE_PREFIX + '-merge';

const TEMP_MERGE_CONFIG_FILE_PREFIX = 'config';
const TEMP_MERGE_CONFIG_FILE_EXTENSION = 'txt';

class TestRunVideoRecorder {
    constructor({ testRun, test, index }, { path, ffmpegPath, encodingOptions }) {
        this.testRun = testRun;
        this.test = test;
        this.index = index;

        this.tempFiles = null;
        this.videoRecorder = null;

        this.path = path;
        this.ffmpegPath = ffmpegPath;
        this.encodingOptions = encodingOptions;
    }

    get testRunInfo() {
        return {
            testIndex: this.index,
            fixture: this.test.fixture.name,
            test: this.test.name,
            alias: this._connection.browserInfo.alias,
            parsedUserAgent: this._connection.browserInfo.parsedUserAgent
        };
    }

    get hasErrors() {
        return !!this.testRun.errs.length;
    }

    get _connection() {
        return this.testRun.browserConnection;
    }

    startCapturing() {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            yield _this.videoRecorder.startCapturing();
        })();
    }

    finishCapturing() {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            yield _this2.videoRecorder.finishCapturing();
        })();
    }

    init() {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            _this3.tempFiles = _this3._generateTempNames();
            _this3.videoRecorder = _this3._createVideoRecorderProcess();

            yield _this3.videoRecorder.init();
        })();
    }

    isVideoSupported() {
        var _this4 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const connectionCapabilities = yield _this4._connection.provider.hasCustomActionForBrowser(_this4._connection.id);

            return connectionCapabilities && connectionCapabilities.hasGetVideoFrameData;
        })();
    }

    _createVideoRecorderProcess() {
        return new _process2.default(this.tempFiles.tempVideoPath, this.ffmpegPath, this._connection, this.encodingOptions);
    }

    _generateTempNames() {
        const id = this._connection.id;

        const tempFileNames = {
            tempVideoPath: `${TEMP_VIDEO_FILE_PREFIX}-${id}.${VIDEO_EXTENSION}`,
            tempMergeConfigPath: `${TEMP_MERGE_CONFIG_FILE_PREFIX}-${id}.${TEMP_MERGE_CONFIG_FILE_EXTENSION}`,
            tmpMergeName: `${TEMP_MERGE_FILE_PREFIX}-${id}.${VIDEO_EXTENSION}`
        };

        for (var _iterator = (0, _entries2.default)(tempFileNames), _isArray = Array.isArray(_iterator), _i = 0, _iterator = _isArray ? _iterator : (0, _getIterator3.default)(_iterator);;) {
            var _ref2;

            if (_isArray) {
                if (_i >= _iterator.length) break;
                _ref2 = _iterator[_i++];
            } else {
                _i = _iterator.next();
                if (_i.done) break;
                _ref2 = _i.value;
            }

            const _ref = _ref2;
            const tempFile = _ref[0];
            const tempName = _ref[1];

            tempFileNames[tempFile] = (0, _path.join)(this.path, tempName);
        }return tempFileNames;
    }
}
exports.default = TestRunVideoRecorder;
module.exports = exports['default'];
