// testcafe/lib/marketing/data-file.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const make_dir_1 = __importDefault(require("make-dir"));
const promisified_functions_1 = require("../utils/promisified-functions");
const render_template_1 = __importDefault(require("../utils/render-template"));
const debug_1 = __importDefault(require("debug"));
const warning_message_1 = __importDefault(require("../notifications/warning-message"));
const MARKETING_DATA_DIR = path_1.default.join(os_1.default.tmpdir(), 'testcafe-marketing');
const MARKETING_DATA_FILE = 'data-file.json';
const DEBUG_LOGGER = debug_1.default('testcafe:marketing');
class DataFile {
    constructor() {
        this.dataDir = MARKETING_DATA_DIR;
        this.dataPath = path_1.default.join(MARKETING_DATA_DIR, MARKETING_DATA_FILE);
    }
    async _exists() {
        try {
            await promisified_functions_1.stat(this.dataPath);
            return true;
        }
        catch (e) {
            return false;
        }
    }
    static _createTemplate() {
        return {
            runCount: 0,
            displayedMessageIndex: void 0
        };
    }
    async _save(data) {
        await promisified_functions_1.writeFile(this.dataPath, JSON.stringify(data));
    }
    async _load() {
        const fileContent = await promisified_functions_1.readFile(this.dataPath);
        return JSON.parse(fileContent);
    }
    async load() {
        try {
            await make_dir_1.default(this.dataDir);
            let content = null;
            if (!await this._exists()) {
                content = DataFile._createTemplate();
                await this._save(content);
            }
            else
                content = await this._load();
            return content;
        }
        catch (e) {
            const message = render_template_1.default(warning_message_1.default.cannotLoadMarketingData, e);
            DEBUG_LOGGER(message);
            return null;
        }
    }
    async save(data) {
        try {
            await this._save(data);
        }
        catch (e) {
            const message = render_template_1.default(warning_message_1.default.cannotSaveMarketingData, e);
            DEBUG_LOGGER(message);
        }
    }
    // NOTE: For testing purposes
    async _remove() {
        try {
            await promisified_functions_1.deleteFile(this.dataPath);
        }
        catch (e) { // eslint-disable-line no-empty
        }
    }
}
exports.default = DataFile;
module.exports = exports.default;
