// testcafe/lib/marketing/data-file.js
'use strict';

exports.__esModule = true;

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _makeDir = require('make-dir');

var _makeDir2 = _interopRequireDefault(_makeDir);

var _promisifiedFunctions = require('../utils/promisified-functions');

var _renderTemplate = require('../utils/render-template');

var _renderTemplate2 = _interopRequireDefault(_renderTemplate);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _warningMessage = require('../notifications/warning-message');

var _warningMessage2 = _interopRequireDefault(_warningMessage);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MARKETING_DATA_DIR = _path2.default.join(_os2.default.tmpdir(), 'testcafe-marketing');
const MARKETING_DATA_FILE = 'data-file.json';

const DEBUG_LOGGER = (0, _debug2.default)('testcafe:marketing');

class DataFile {
    constructor() {
        this.dataDir = MARKETING_DATA_DIR;
        this.dataPath = _path2.default.join(MARKETING_DATA_DIR, MARKETING_DATA_FILE);
    }

    _exists() {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            try {
                yield (0, _promisifiedFunctions.stat)(_this.dataPath);

                return true;
            } catch (e) {
                return false;
            }
        })();
    }

    static _createTemplate() {
        return {
            runCount: 0,
            displayedMessageIndex: 0
        };
    }

    _save(data) {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            yield (0, _promisifiedFunctions.writeFile)(_this2.dataPath, (0, _stringify2.default)(data));
        })();
    }

    _load() {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const fileContent = yield (0, _promisifiedFunctions.readFile)(_this3.dataPath);

            return JSON.parse(fileContent);
        })();
    }

    load() {
        var _this4 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            try {
                yield (0, _makeDir2.default)(_this4.dataDir);

                let content = null;

                if (!(yield _this4._exists())) {
                    content = DataFile._createTemplate();

                    yield _this4._save(content);
                } else content = yield _this4._load();

                return content;
            } catch (e) {
                const message = (0, _renderTemplate2.default)(_warningMessage2.default.cannotLoadMarketingData, e);

                DEBUG_LOGGER(message);

                return null;
            }
        })();
    }

    save(data) {
        var _this5 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            try {
                yield _this5._save(data);
            } catch (e) {
                const message = (0, _renderTemplate2.default)(_warningMessage2.default.cannotSaveMarketingData, e);

                DEBUG_LOGGER(message);
            }
        })();
    }

    // NOTE: For testing purposes
    _remove() {
        var _this6 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            try {
                yield (0, _promisifiedFunctions.deleteFile)(_this6.dataPath);
            } catch (e) {// eslint-disable-line no-empty
            }
        })();
    }
}
exports.default = DataFile;
module.exports = exports['default'];
