// testcafe/lib/api/request-hooks/hook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const testcafe_hammerhead_1 = require("testcafe-hammerhead");
const lodash_1 = require("lodash");
class RequestHook {
    constructor(requestFilterRules, responseEventConfigureOpts) {
        this.requestFilterRules = this._prepareRequestFilterRules(requestFilterRules);
        this._instantiatedRequestFilterRules = [];
        this.responseEventConfigureOpts = responseEventConfigureOpts;
        this.warningLog = null;
    }
    _prepareRequestFilterRules(rules) {
        if (rules)
            return lodash_1.castArray(rules);
        return [testcafe_hammerhead_1.RequestFilterRule.ANY];
    }
    _instantiateRequestFilterRules() {
        this._instantiatedRequestFilterRules = [];
        this.requestFilterRules.forEach(rule => {
            const instantiatedRule = rule instanceof testcafe_hammerhead_1.RequestFilterRule ? rule : new testcafe_hammerhead_1.RequestFilterRule(rule);
            this._instantiatedRequestFilterRules.push(instantiatedRule);
        });
    }
    async onRequest( /*RequestEvent event*/) {
        throw new Error('Not implemented');
    }
    _onConfigureResponse(event) {
        if (!this.responseEventConfigureOpts)
            return;
        event.opts.includeHeaders = this.responseEventConfigureOpts.includeHeaders;
        event.opts.includeBody = this.responseEventConfigureOpts.includeBody;
    }
    async onResponse( /*ResponseEvent event*/) {
        throw new Error('Not implemented');
    }
}
exports.default = RequestHook;
module.exports = exports.default;
